### ISSUE TITLE
Inferred type not valid

### ISSUE THREAD

**Issue by weaversa** [GaloisInc/cryptol#255]
Here's my function

```
widen c = (0#c)
```

Cryptol happily accepts this. I can ask Cryptol for the type:

```
Main> :t widen
widen : {a, b} (fin b) => [a] -> [a + b]
```

Adding the type to the definition in the file causes problems.

```
widen : {a, b} (fin b) => [a] -> [a + b]
widen c = (0#c)
```

```
Main> :r
Loading module Cryptol
Loading module Main

[error] at ./a.cry:49:1--49:16:
  Failed to validate user-specified signature.
    In the definition of 'widen', at ./a.cry:49:1--49:6:
      for any type a, b
        fin b
      =>
      a + b == ?r18 + a
        arising from
        matching types
        at ./a.cry:49:12--49:15
      fin ?r18
        arising from
        use of expression (#)
        at ./a.cry:49:12--49:15
  where
  ?r18 is type parameter 'bits'
          of literal or demoted expression
          at ./a.cry:49:12--49:15
```

**Comment by robdockins** [comment]
After a little looking into this, the issue strikes me as a pretty subtle one.  The problem, roughly, is that the type inference algorithm has named the implicit numeric width of the constant `0` with the name `b`.  However, when we go back to type check with the resulting signature, nothing explicitly ties the value of the type parameter `b` to the constant `0`.  In particular, in the case where `a = inf`, the size of `b` is totally disconnected from the width of the `0` constant.  For example, consider the following:

```
widen : {b} (fin b) => [inf] -> [inf + b]
widen c = ( 0 # c )
```

Cryptol will accept this, but will use defaulting to choose the width of `0`, because it is unconstrained.
```
[warning] at ./asdf.cry:6:1--6:20:
  Defaulting type parameter 'bits'
             of literal or demoted expression
             at ./asdf.cry:6:13--6:18
  to 0
```

On the other hand, if we force `a` to be finite, then `b` is uniquely determined (because then addition satisfies cancellation laws).  Cryptol accepts the following without complaint.
```
widen : {b} (fin a, fin b) => [a] -> [a + b]
widen c = ( 0 # c )
```

Finally, if we add an ascription to the term, we can directly connect the type variable to the width of the constant, as we wanted in the first place.
```
widen : {a, b} (fin b) => [a] -> [a + b]
widen c = ( (0:[b]) # c )
```

I'm not sure what to do about this, except perhaps to document the fact that sometimes Cryptol will infer types that cannot be directly pasted in as a signature without modifying the original term to add some type ascriptions.  

The other alternatives that come to mind seem worse (e.g., making it so the original term doesn't typecheck; or inferring a type which is not most general; getting rid of `inf`).

**Comment by jpziegler** [comment]
How about printing out a warning that the inferred type is not "copy-pastable"?

If you can do that, then maybe also print something (such as a line and column number) that indicates where each type variable is coming from in order to help guide the user to make the changes necessary to use this explicit type with their definition. (Since that's generally something to be encouraged.)

Best yet: something that will make the changes to your definition for you. Here's dream mock-up! (Forgive the misuse of terminology.)
```
Main> :t widen
widen : {a, b} (fin b) => [a] -> [a + b]
[warning] implicit type includes type variables referencing the definition body.
  Use ":definition" for a definition that will properly check against this type.
Main> :definition widen
widen : {a, b} (fin b) => [a] -> [a + b]
widen c = ( (0:[b]) # c )
Main>
```
Probably a tall order.

**Comment by robdockins** [comment]
I discussed this issue with @yav recently.  He noted that checking for this situation (inferred type cannot be directly used as a signature) is similar to the type ambiguity check as done in GHC, and that we could perhaps perform a similar check here.

We also discussed the possibility of implementing something like the `:definition` command suggested above by adding type ascriptions to all the subterms of a term (which _\*handwaving\*_ should be sufficient to make the term check with the inferred signature), and then progressively removing those that are not necessary for the term to typecheck.
